### DISCUSSION TITLE
Returning a table on button click

### DISCUSSION THREAD

**Discussion by rohitsathish** [zauberzeug/nicegui#2098] [Q&A]
### Question

I would like to return a ```ui.table``` component when a button is clicked. 

If possible I'd like to not instantiate an empty table before clicking the button, but I don't think thats possible. Also it looks like ```ui.table``` doesn't support an empty instantiation. So is there a way to simply insert a new table with a button click? I think I'm missing something obvious here.

Here's my approach so far - this seems to add new tables, not modify the existing one.:

```py
import pandas as pd
from nicegui import ui

@ui.page("/")
def main_page():
    ui.add_head_html(
        f"<style>{(Path(__file__).parent.parent / 'static' / 'style.css').read_text()}</style>"
    )

    df_show = ui.table(columns = [1], rows = [1])

    async def load_df():
        df = pd.DataFrame({'a': ['c'], 'b': ['d']}, index=['1', '2'])
        df_show = ui.table.from_pandas(df)

    ui.button('Load Dataframe', on_click=load_df)
    

ui.run(uvicorn_reload_includes="*.py, *.html, *.js, *.css")
```

**Comment by natankeddem** [comment]
You could create a container (`ui.row`, `ui.column`, ect...). Then later on you can insert any elements in that container when you want. For example.
```py
from nicegui import ui

row = ui.row()

def insert_table_into_row():
    with row:
        ui.table(columns=[1], rows=[1])

ui.button("Insert table into the row.", on_click=insert_table_into_row)
ui.run()
```

**Reply by rohitsathish** [reply]
This is definitely a cleaner approach than what I had, but each time I press the button, it adds a new table. I would only like to add the table if it doesn't exist.

My current unclean approach is:

```
df = None

@ui.refreshable
async def df_ui():
    if df is not None:
        ui.table.from_pandas(df)

async def load_df():
    nonlocal df
    df = pd.DataFrame({'a': ['c'], 'b': ['d']}, index=['1', '2'])
    df_ui.refresh()

ui.button('Load Dataframe', on_click=load_df)   
df_ui()
```

**Reply by natankeddem** [reply]
If you can only deal with `pandas` you will have to use that approach I think. If you can deal directly with the data you can just update fields descretely.
```py
from nicegui import ui
import pandas as pd
import random

table = None

def just_table_stuff():
    global table
    if table is None:
        table = ui.table.from_pandas(pd.DataFrame({"a": ["c"], "b": ["d"]}, index=["1", "2"]))
    else:
        table.add_rows({"a": "e", "b": random.randint(0, 9)})

ui.button("Load Dataframe or Add Rows", on_click=just_table_stuff)

ui.run()
```

**Reply by rohitsathish** [reply]
Yup, I think you've got the right idea. Made some minor modifications to have a state dictionary so I don't have a bunch of instantiations. 

```python
from nicegui import ui
import pandas as pd

state = {}

@ui.page("/")
def main_page():

    def df_ui(df):
        if 'table' not in state:
            state['table'] = ui.table.from_pandas(df)
            
    ui.button("Load Dataframe", on_click=lambda: df_ui(pd.DataFrame({"a": ["c"], "b": ["d"]}, index=["1","2"])))

ui.run(uvicorn_reload_includes="*.py, *.html, *.js, *.css")

**Comment by falkoschindler** [comment]
I would do something like this:

- Create a table with empty rows and columns.
- Hide it.
- When clicking the button, update rows and columns and show table.

```py
table = ui.table(columns=[], rows=[])
table.visible = False

def load_df():
    df = pd.DataFrame({'a': ['c'], 'b': ['d']}, index=['1', '2'])
    table.columns = [{'name': col, 'label': col, 'field': col} for col in df.columns]
    table.rows = df.to_dict('records')
    table.visible = True

ui.button('Load Dataframe', on_click=load_df)
```

The code from converting the dataframe to rows and columns is borrowed from
https://github.com/zauberzeug/nicegui/blob/a7bf8d2b6f4662a0e31c11599d5598210f8e3818/nicegui/elements/table.py#L101-L102
